### PR TITLE
Initialize Name earlier in the Create process: GenerateName

### PR DESCRIPTION
**Pull request by aojea**
/kind bug
/kind cleanup

```release-note
NONE
```

Following up on `Initialize UID earlier in the Create process` #110646 

Before:
Create()
BeginCreate()
BeforeCreate()
GenerateName() if Name empty and len(GenerateName) > 0 <---------------------
strategy code

After:
Create()
GenerateName() if Name empty and len(GenerateName) > 0 <---------------------
BeginCreate()
BeforeCreate()
strategy code

Some comments from the slack discussion https://kubernetes.slack.com/archives/C0EG7JC6T/p1661439190979379

> thockin:  our REST semantics don't allow for unnamed resources and I don't think we want to make naming even more resource-centric

> liggit: moving it would make PrepareForCreate unable to tell if the name was user-specified or generated, and would remove the ability of PrepareForCreate to influence the generated name by modifying metadata.generateName
moving it would make PrepareForCreate unable to tell if the name was user-specified or generated, and would remove the ability of PrepareForCreate to influence the generated name by modifying metadata.generateName
> but I don't see any PrepareForCreate implementations that do anything like that, fortunately
> and the NameGenerator interface only takes metadata.generateName as the input param, so no name generation can directly depend on any other data in the object

Jordan expressed some possible issues

> and would this relocation change what happens if a generated name gets a conflict? if we would retry the generation internally previously, we'd want to make sure we'd still retry (I don't actually know if we automatically retry internally or not)

I've added a test to check the behavior before and after and it always pass

### CODE REVIEW

**Comment by aojea** [comment]
wow, lot's of test failing now 

> k8s.io/kubernetes/vendor/k8s.io/apiextensions-apiserver/test/integration: TestInvalidCRUD expand_more

it seems all tests were using an empty name?

>     resttest.go:482: Expected 'does not match the namespace sent on the request' error, got 'Internal error occurred: metadata.name was not set'

**Comment by aojea** [comment]
the failures are legit, but I need guidance:
- is the approach correct?
- are the test wrong?
- ...

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go [hunk 1]
```diff
@@ -105,6 +105,11 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx context.Context, obj runtime.
 		return errors.NewInternalError(fmt.Errorf("system metadata was not initialized"))
 	}
 
+	// ensure the name has been set
+	if len(objectMeta.GetName()) == 0 {
+		return errors.NewInternalError(fmt.Errorf("metadata.name was not set"))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:110]: internal error here isn't right... a missing name should continue and be caught in validation and result in a Required validation error

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:110]: if you want to make sure generation happened, make this conditional on `len(objectMeta.GetGenerateName()) > 0`

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go]: yeah,

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 2]
```diff
@@ -570,6 +573,9 @@ func (e *Store) Update(ctx context.Context, name string, objInfo rest.UpdatedObj
 				return nil, nil, err
 			} else {
 				rest.FillObjectMetaSystemFields(objectMeta)
+				if len(objectMeta.GetGenerateName()) > 0 && len(objectMeta.GetName()) == 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:576]: is it even possible for name to be len 0 here? update requires the name be pre-specified... populating name from generateName in update does not seem correct to me

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:576]: ... this area of code is not my thing as you can see :smile:

Comment (lavalamp) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:576]: Yeah, update supports creating but you have to know the name.

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:576]:
```
kubectl -v7 apply -f pod.yaml                           
I0830 22:49:24.561890 4007126 loader.go:372] Config loaded from file:  /tmp/kind.conf
I0830 22:49:24.562775 4007126 round_trippers.go:463] GET https://127.0.0.1:35399/openapi/v2?timeout=32s
I0830 22:49:24.562784 4007126 round_trippers.go:469] Request Headers:
I0830 22:49:24.562802 4007126 round_trippers.go:473]     Accept: application/com.github.proto-openapi.spec.v2@v1.0+protobuf
I0830 22:49:24.562808 4007126 round_trippers.go:473]     User-Agent: kubectl/v1.23.4 (linux/amd64) kubernetes/e6c093d
I0830 22:49:24.598770 4007126 round_trippers.go:574] Response Status: 200 OK in 35 milliseconds
F0830 22:49:24.738982 4007126 helpers.go:118] error: from gen-: cannot use generate name with apply
```

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:576]: hmm, this test fails, but I don't know if there is some check on the name before calling Update

```diff
+++ b/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store_test.go
@@ -1070,6 +1070,34 @@ func TestStoreCreateOnUpdateHooks(t *testing.T) {
        }
 }
 
+func TestCreateOnUpdateGenerateName(t *testing.T) {
+       pod := &example.Pod{
+               ObjectMeta: metav1.ObjectMeta{GenerateName: "foo-", Namespace: "test"},
+               Spec:       example.PodSpec{NodeName: "machine"},
+       }
+
+       testContext := genericapirequest.WithNamespace(genericapirequest.NewContext(), "test")
+       destroyFunc, registry := NewTestGenericStoreRegistry(t)
+       defer destroyFunc()
+
+       obj, ok, err := registry.Update(testContext, "", rest.DefaultUpdatedObjectInfo(pod), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc, true, &metav1.UpdateOptions{})
+       if err != nil {
+               t.Fatalf("Unexpected error: %v", err)
+       }
+       if !ok {
+               t.Fatalf("Expected object to be created")
+       }
+
+       accessor, err := meta.Accessor(obj)
+       if err != nil {
+               t.Errorf("Unexpected error: %v", err)
+       }
+
+       if !strings.HasPrefix(accessor.GetName(), "foo-") {
+               t.Fatalf("Expected object name to be generated from prefix foo-")
+       }
+}
+
```

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:576]: ic, that is only valid for the test, you have to specify the name as part of the URL to call an Update

File: staging/src/k8s.io/apiserver/pkg/registry/rest/create.go [hunk 3]
```diff
@@ -105,6 +105,11 @@ func BeforeCreate(strategy RESTCreateStrategy, ctx context.Context, obj runtime.
 		return errors.NewInternalError(fmt.Errorf("system metadata was not initialized"))
 	}
 
+	// ensure the name has been generated
+	if len(objectMeta.GetGenerateName()) > 0 && len(objectMeta.GetName()) == 0 {
+		return errors.NewInternalError(fmt.Errorf("metadata.name was not generated"))
+	}
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:111]: @liggitt @carlory should this be more strict and require that the name exists here no matter what?

is there any reason at this point an object may not have the name? (ref https://github.com/kubernetes/kubernetes/pull/135364 )

Comment (carlory) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:111]: @aojea the `TokenRequest` resource doesn't require a name.

https://github.com/carlory/kubernetes/blob/db6805f515b047a04fd5d2f679726bf0331998d8/pkg/apis/authentication/validation/validation.go#L29
https://github.com/carlory/kubernetes/blob/db6805f515b047a04fd5d2f679726bf0331998d8/pkg/registry/authentication/tokenreview/storage.go#L69

Comment (carlory) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:111]: This resource does not need to be stored in etcd.

Comment (aojea) [reply staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:111]: good catch

**Pull request merged**